tensorflow/lite/micro/tools/requantize_flatbuffer_utils.py

```diff
@@ -0,0 +1,215 @@
+# Copyright 2023 The TensorFlow Authors. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# =============================================================================
+import numpy as np
+from absl import logging
+
+MIN_INT8, MAX_INT8 = -128, 128  # Wide
+
+MIN_INT16, MAX_INT16 = -32767, 32767  # Narrow range
+
+# Map flatbuffer tensor type code to numpy data type. see Table TensorType in tensorflow/lite/schema/schema.fbs
+TENSOR_CODE_TYPE = {
+    0: np.float32,
+    1: np.float16,
+    2: np.int32,
+    3: np.uint8,
+    4: np.int64,
+    5: np.string_,
+    6: np.bool_,
+    7: np.int16,
+    8: np.complex64,
+    9: np.int8,
+    10: np.float64,
+    11: np.complex128,
+    12: np.uint64,
+    13: "RESORCE",
+    14: "VARIANT",
+    15: "UINT32",
+    16: "UINT16",
+    17: "INT4",
+}
+
+TENSOR_TYPE_CODE = dict((reversed(item) for item in TENSOR_CODE_TYPE.items()))
+"""
+Quantization utils functions
+"""
+
+
+def clip_range(vals, bit_width):
+  """Mimic integer calculation.
+
+  Clip the range of vals based on bit width.
+
+  e.g., clip_range([300], 8) = [127] since int8 have range [-128, 127]
+
+  Args:
+      vals (np.array): float representation of the integer values
+      bit_width (int): number of desired bits for vals
+
+  Returns:
+      np.array : clipped vals
+  """
+  # Numpy integer calculation does not do saturation. Implement here
+  half = np.power(2.0, bit_width) / 2
+  min_val, max_val = -half, half - 1
+  if vals.max() > max_val or vals.min() < min_val:
+    logging.info(f"WARNING: integer overflow!")
```

Comment:
nit: don't we expect this, since the function is called clip? I wouldn't put a log in here.

Reply:
I found this information very useful because integer overflow is always the silent killer that lowers the overall model precision.